src/VisualStudio/Core/Def/DocumentOutline/DocumentSymbolDataViewModel.cs

```diff
@@ -0,0 +1,114 @@
+        /// Typically, this is the identifier name for the symbol
+        /// </summary>
+        public SnapshotSpan SelectionRangeSpan => _data.SelectionRangeSpan;
+        public SymbolKind SymbolKind => _data.SymbolKind;
+        public ImageMoniker ImageMoniker => _data.SymbolKind.GetImageMoniker();
+
+        private bool _isExpanded;
+        public bool IsExpanded
+        {
+            get => _isExpanded;
+            set => SetProperty(ref _isExpanded, value);
+        }
+
+        private bool _isSelected;
+
+        public bool IsSelected
+        {
+            get => _isSelected;
+            set => SetProperty(ref _isSelected, value);
+        }
+
+        public DocumentSymbolDataViewModel(
+            DocumentSymbolData data,
+            ImmutableArray<DocumentSymbolDataViewModel> children,
+            bool isExpanded,
+            bool isSelected)
+        {
+            _data = data;
+            Children = children;
+            _isExpanded = isExpanded;
+            _isSelected = isSelected;
+        }
+
+        private static readonly PropertyChangedEventArgs _isExpandedPropertyChangedEventArgs = new PropertyChangedEventArgs(nameof(IsExpanded));
+        private static readonly PropertyChangedEventArgs _isSelectedPropertyChangedEventArgs = new PropertyChangedEventArgs(nameof(IsSelected));
+
+        private void NotifyPropertyChanged([CallerMemberName] string? propertyName = null)
+            => PropertyChanged?.Invoke(this, propertyName switch
+            {
+                nameof(IsExpanded) => _isExpandedPropertyChangedEventArgs,
+                nameof(IsSelected) => _isSelectedPropertyChangedEventArgs,
+                _ => new PropertyChangedEventArgs(propertyName)
+            });
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        private void SetProperty(ref bool field, bool value, [CallerMemberName] string propertyName = "")
+        {
+            // Note: we do not lock here. Worst case is that we fire multiple
+            //       NotifyPropertyChanged events which WPF can handle.
+            if (field == value)
+            {
+                return;
+            }
+
+            field = value;
+            NotifyPropertyChanged(propertyName);
+        }
+
+        public static bool operator ==(DocumentSymbolDataViewModel left, DocumentSymbolDataViewModel right)
+            => left.Equals(right);
+
+        public static bool operator !=(DocumentSymbolDataViewModel left, DocumentSymbolDataViewModel right)
+            => !(left == right);
+
+        public override bool Equals(object obj)
+            => Equals(obj as DocumentSymbolDataViewModel);
+
+        public bool Equals(DocumentSymbolDataViewModel? other)
+            => (object)this == other ||
+                 (other is not null &&
+                   (RangeSpan.Span, Name, SymbolKind) ==
+                   (other.RangeSpan.Span, other.Name, other.SymbolKind));
```

Comment:
this generally feels weird.  you could have different items with different snapshots... and we'd think they were the same?  intentiona?

Reply:
yes, we don't want to redraw if its an identical symbol even if the snapshot is different

Reply:
so what we do in tagging, is we normalize to the same snapshot (either move the old guys forward, or the new guys backward) and we validate they're truly the same.

The above, for example, will redraw, if user just adds a newline to the file (or even just types a space).  However, conceptually, you will get all the same items (just shifted).  So we want to update our data in that case, but not redraw the UI.